File: tests/string.c
#include "tests.h"
#include <string.h>

#define START_TEST(t) \
    diag(#t);         \
    test_##t();

void test_strtok()
{
    char str[] = "- This, a sample string.";
    char* pch;
    printf("Splitting string \"%s\" into tokens:\n", str);
    pch = strtok(str, " ,.-");
    printf("strtok: `%s`\n", pch);
    // while (pch != NULL)
    // {
    // printf ("%s\n",pch);
    // pch = strtok (NULL, " ,.-");
    // }
}

void test_strncmp()
{
    char str[][5] = { "R2D2", "C3PO", "R2A6" };
    int n;
    puts("Looking for R2 astromech droids...");
    for (n = 0; n < 3; n++) {
        if (strncmp(str[n], "R2xx", 2) == 0) {
            printf("found %s\n", str[n]);
        }
    }
}

void test_strcspn()
{
    char str[] = "fcba73";
    char keys[] = "1234567890";
    int i;
    i = strcspn(str, keys);
    printf("The first number in str is at position %d.\n", i + 1);
}

void test_strpbrk()
{
    char str[] = "This is a sample string";
    char key[] = "aeiou";
    char* pch;
    printf("Vowels in '%s': ", str);
    pch = strpbrk(str, key);
    printf("strpbrk: `%s`\n", pch);
    // while (pch != NULL) {
    // printf ("%c " , *pch);
    // pch = strpbrk (pch+1,key);
    // }
    printf("\n");
}

void test_strspn()
{
    int i;
    char strtext[] = "129th";
    char cset[] = "1234567890";
    i = strspn(strtext, cset);
    printf("The initial number has %d digits.\n", i);
}

int main()
{
    plan(44);

    diag("TODO: __builtin_object_size");
    // https://github.com/Konstantin8105/c4go/issues/359

    {
        diag("strcpy");
        char* src = "foo bar\0baz";
        char dest1[40];
        char* dest2;
        dest2 = strcpy(dest1, src);
        is_streq(dest1, "foo bar");
        is_streq(dest2, "foo bar");
    }

    diag("strncpy");

    // If the end of the source C string (which is signaled by a null-character)
    // is found before num characters have been copied, destination is padded
    // with zeros until a total of num characters have been written to it.
    {
        char* src = "foo bar\0baz";
        char dest1[40];
        char* dest2;

        dest1[7] = 'a';
        dest1[15] = 'b';
        dest1[25] = 'c';
        dest2 = strncpy(dest1, src, 20);

        is_eq(dest1[0], 'f');
        is_eq(dest1[7], 0);
        is_eq(dest1[15], 0);
        is_eq(dest1[25], 'c');

        is_eq(dest2[0], 'f');
        is_eq(dest2[7], 0);
        is_eq(dest2[15], 0);
        is_eq(dest2[25], 'c');

        is_streq(dest1, "foo bar");
        is_streq(dest2, "foo bar");
    }

    // No null-character is implicitly appended at the end of destination if
    // source is longer than num. Thus, in this case, destination shall not be
    // considered a null terminated C string (reading it as such would
    // overflow).
    {
        char* src = "foo bar\0baz";
        char dest1[40];
        char* dest2;

        dest1[7] = 'a';
        dest1[15] = 'b';
        dest1[25] = 'c';
        dest2 = strncpy(dest1, src, 5);

        is_eq(dest1[0], 'f');
        is_eq(dest1[7], 'a');
        is_eq(dest1[15], 'b');
        is_eq(dest1[25], 'c');

        is_eq(dest2[0], 'f');
        is_eq(dest2[7], 'a');
        is_eq(dest2[15], 'b');
        is_eq(dest2[25], 'c');
    }

    {
        diag("strlen");
        is_eq(strlen(""), 0);
        is_eq(strlen("a"), 1);
        is_eq(strlen("foo"), 3);
        // NULL causes a seg fault.
        // is_eq(strlen(NULL), 0);
        is_eq(strlen("fo\0o"), 2);
    }
    {
        diag("strcat");
        char str[80];
        for (int i = 0; i < 80; i++)
            str[i] = 0;
        strcpy(str, "these ");
        strcat(str, "strings ");
        strcat(str, "are ");
        strcat(str, "concatenated.");
        is_streq(str, "these strings are concatenated.");
    }
    {
        diag("strncat");
        char str[80];
        for (int i = 0; i < 80; i++)
            str[i] = 0;
        strncpy(str, "these", 3);
        strncat(str, " strings", 7);
        strncat(str, " is", 3);
        strncat(str, " concatenated.", 14);
        is_streq(str, "the string is concatenated.");
    }
    {
        diag("strcmp");
        {
            char* a = "ab";
            char* b = "ab";
            is_true(strcmp(a, b) == 0);
        }
        {
            char* a = "bb";
            char* b = "ab";
            is_true(strcmp(a, b) > 0);
        }
        {
            char* a = "ab";
            char* b = "bb";
            is_true(strcmp(a, b) < 0);
        }
    }
    {
        diag("strchr");
        char str[] = "This is a sample string";
        char* pch;
        int amount = 0;
        pch = strchr(str, 's');
        while (pch != NULL) {
            pch = strchr(pch + 1, 's');
            amount++;
        }
        is_eq(amount, 4);
    }
    {
        diag("memset");
        char str[] = "almost every programmer should know memset!";
        char* ptr = memset(str, '-', 6);
        is_streq(str, "------ every programmer should know memset!");
        is_eq(ptr - str, 0);
    }
    {
        diag("memmove");
        char str[] = "memmove can be very useful......";
        memmove(str + 20, str + 15, 11);
        is_streq(str, "memmove can be very very useful.");
    }
    {
        diag("memcmp");
        char a1[] = { 'a', 'b', 'c' };
        char a2[] = "abd";
        is_true(memcmp(a1, a1, 3) == 0);
        is_true(memcmp(a1, a2, 3) < 0);
        is_true(memcmp(a2, a1, 3) > 0);
    }
    {
        diag("strstr");
        char str[] = "one two three";

        printf("%s\n", strstr(str, "one"));
        is_streq(strstr(str, "one"), "one two three");

        printf("%s\n", strstr(str, "two"));
        is_streq(strstr(str, "two"), "two three");

        printf("%s\n", strstr(str, "three"));
        is_streq(strstr(str, "three"), "three");
        is_null(strstr(str, "fo"));
    }
    {
        diag("memcpy");
        char myname[] = "Pierre de Fermat";
        char name[40];
        memcpy(name, myname, strlen(myname) + 1);
        is_streq(name, myname);
        printf("name = `%s`\n", name);
    }
    {
        diag("strrchr");
        char str[] = "This is a sample string";
        char failData[] = "faildata";
        char* pch = &failData;
        pch = strrchr(str, 's');
        is_eq(pch - str + 1, 18);
    }
    {
        diag("strdup");
        const char* s1 = "String";
        char* s2 = strdup(s1);
        is_streq(s1, s2);
        is_true(s1 != s2);
    }
    START_TEST(strtok);
    START_TEST(strncmp);
    START_TEST(strcspn);
    START_TEST(strpbrk);
    START_TEST(strspn);
    {
        diag("strerror");
        strerror(0);
    }

    done_testing();
}

File: noarch/string.go
package noarch

import (
	"bytes"
	"unsafe"
)

// Strlen returns the length of a string.
//
// The length of a C string is determined by the terminating null-character: A
// C string is as long as the number of characters between the beginning of the
// string and the terminating null character (without including the terminating
// null character itself).
func Strlen(a []byte) int32 {
	// TODO: The transpiler should have a syntax that means this proxy function
	// does not need to exist.

	return int32(len(CStringToString(a)))
}

// Strcpy copies the C string pointed by source into the array pointed by
// destination, including the terminating null character (and stopping at that
// point).
//
// To avoid overflows, the size of the array pointed by destination shall be
// long enough to contain the same C string as source (including the terminating
// null character), and should not overlap in memory with source.
func Strcpy(dest, src []byte) []byte {
	for i, c := range src {
		dest[i] = c

		// We only need to copy until the first NULL byte. Make sure we also
		// include that NULL byte on the end.
		if c == '\x00' {
			break
		}
	}

	return dest
}

// Strncpy copies the first num characters of source to destination. If the end
// of the source C string (which is signaled by a null-character) is found
// before num characters have been copied, destination is padded with zeros
// until a total of num characters have been written to it.
//
// No null-character is implicitly appended at the end of destination if source
// is longer than num. Thus, in this case, destination shall not be considered a
// null terminated C string (reading it as such would overflow).
//
// destination and source shall not overlap (see memmove for a safer alternative
// when overlapping).
func Strncpy(dest, src []byte, len32 int32) []byte {
	len := int(len32)
	// Copy up to the len or first NULL bytes - whichever comes first.
	i := 0
	for ; i < len && src[i] != 0; i++ {
		dest[i] = src[i]
	}

	// The rest of the dest will be padded with zeros to the len.
	for ; i < len; i++ {
		dest[i] = 0
	}

	return dest
}

// Strcat - concatenate strings
// Appends a copy of the source string to the destination string.
// The terminating null character in destination is overwritten by the first
// character of source, and a null-character is included at the end
// of the new string formed by the concatenation of both in destination.
func Strcat(dest, src []byte) []byte {
	Strcpy(dest[Strlen(dest):], src)
	return dest
}

// Strncat - concatenate strings
// Appends at most count characters of the source string to the destination string.
// The terminating null character in destination is overwritten by the first
// character of source, and a null-character is included at the end
// of the new string formed by the concatenation of both in destination.
func Strncat(dest, src []byte, len int32) []byte {
	Strncpy(dest[Strlen(dest):], src, len)
	return dest
}

// Strcmp - compare two strings
// Compares the C string str1 to the C string str2.
func Strcmp(str1, str2 []byte) int32 {
	return int32(bytes.Compare([]byte(CStringToString(str1)), []byte(CStringToString(str2))))
}

// Strchr - Locate first occurrence of character in string
// See: http://www.cplusplus.com/reference/cstring/strchr/
func Strchr(str []byte, ch32 int32) []byte {
	ch := int(ch32)
	i := 0
	for {
		if str[i] == '\x00' {
			break
		}
		if int(str[i]) == ch {
			return str[i:]
		}
		i++
	}
	return nil
}

// Strstr finds the first occurrence of the null-terminated byte string
// pointed to by substr in the null-terminated byte string pointed to by str.
// The terminating null characters are not compared.
func Strstr(str, subStr []byte) []byte {
	if subStr == nil {
		return str
	}
	if subStr[0] == '\x00' {
		return str
	}

	k := 0
	for i := range subStr {
		if subStr[i] == '\x00' {
			k = i
			break
		}
	}

	index := bytes.Index(str, subStr[:k])
	if index < 0 {
		return nil
	}
	return str[index:]
}

// Memset sets the first num bytes of the block of memory pointed by ptr to
// the specified value (interpreted as an unsigned char)
func Memset(ptr []byte, value byte, num uint32) []byte {
	for i := 0; uint32(i) < num; i++ {
		ptr[i] = value
	}
	return ptr
}

// Memmove move block of memory
func Memmove(ptr, src interface{}, num uint32) interface{} {
	// both types is []byte
	if p, ok := ptr.([]byte); ok {
		if s, ok := src.([]byte); ok {
			for i := int(num); i >= 0; i-- {
				if i >= len(s) {
					continue
				}
				if i >= len(p) {
					continue
				}
				p[i] = s[i]
				if s[i] == '\x00' {
					break
				}
			}
			return s
		}
	}

	p1 := (*[100000]byte)(unsafe.Pointer(uintptr(unsafe.Pointer(&ptr))))
	p2 := (*[100000]byte)(unsafe.Pointer(uintptr(unsafe.Pointer(&src))))
	for i := int(num); i >= 0; i-- {
		p1[i] = p2[i]
	}
	return p1
}

// Memcmp - compare two buffers
// Compares the first count characters of the objects pointed to be lhs and rhs
func Memcmp(lhs []byte, rhs []byte, count uint32) int32 {
	for i := 0; uint32(i) < count; i++ {
		if int(lhs[i]) < int(rhs[i]) {
			return -1
		} else if int(lhs[i]) > int(rhs[i]) {
			return 1
		}
	}
	return 0
}

func Strrchr(source []byte, c32 int32) []byte {
	c := int(c32)
	ch := byte(c)
	pos := len(source)
	for i := range source {
		if source[i] == '\x00' {
			pos = i
			break
		}
	}
	for i := pos; i >= 0; i-- {
		if source[i] == ch {
			return source[i:]
		}
	}
	return source
}

func Strdup(s []byte) (rs []byte) {
	for i := range s {
		if s[i] == '\x00' {
			break
		}
		rs = append(rs, s[i])
	}
	return
}

func Strerror(e int32) []byte {
	return []byte("strerror")
}
